src/mlpack/methods/kde/kde_model.hpp

```diff
@@ -0,0 +1,381 @@
+ */
+class KernelNormalizer
+{
+ private:
+  // SFINAE helper to check if has a Normalizer function.
+  HAS_MEM_FUNC(Normalizer, HasNormalizer);
+
+ public:
+  //! Normalization not needed.
+  template<typename KernelType>
+  static void ApplyNormalizer(
+      KernelType& /* kernel */,
+      const size_t /* dimension */,
+      arma::vec& /* estimations */,
+      const typename std::enable_if<
+          !HasNormalizer<KernelType, double(KernelType::*)(size_t)>::value>::
+          type* = 0)
+  { return; }
+
+  //! Normalize kernels that have normalizer.
+  template<typename KernelType>
+  static void ApplyNormalizer(
+      KernelType& kernel,
+      const size_t dimension,
+      arma::vec& estimations,
+      const typename std::enable_if<
+          HasNormalizer<KernelType, double(KernelType::*)(size_t)>::value>::
+          type* = 0)
+  {
+    estimations /= kernel.Normalizer(dimension);
+  }
+};
+
+/**
+ * DualMonoKDE computes a Kernel Density Estimation on the given KDEType.
+ * It performs a monochromatic KDE.
+ */
+class DualMonoKDE : public boost::static_visitor<void>
+{
+ private:
+  //! Vector to store the KDE results.
+  arma::vec& estimations;
+
+ public:
+  //! Alias template necessary for visual C++ compiler.
+  template<typename KernelType,
+           template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  using KDETypeT = KDEType<KernelType, TreeType>;
+
+  //! Default DualMonoKDE on some KDEType.
+  template<typename KernelType,
+           template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  void operator()(KDETypeT<KernelType, TreeType>* kde) const;
+
+  // TODO Implement specific cases where a leaf size can be selected.
+
+  //! DualMonoKDE constructor.
+  DualMonoKDE(arma::vec& estimations);
+};
+
+/**
+ * DualBiKDE computes a Kernel Density Estimation on the given KDEType.
+ * It performs a bichromatic KDE.
+ */
+class DualBiKDE : public boost::static_visitor<void>
+{
+ private:
+  //! Query set dimensionality.
+  const size_t dimension;
+
+  //! The query set for the KDE.
+  const arma::mat& querySet;
+
+  //! Vector to store the KDE results.
+  arma::vec& estimations;
+
+ public:
+  //! Alias template necessary for visual C++ compiler.
+  template<typename KernelType,
+           template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  using KDETypeT = KDEType<KernelType, TreeType>;
+
+  //! Default DualBiKDE on some KDEType.
+  template<typename KernelType,
+           template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  void operator()(KDETypeT<KernelType, TreeType>* kde) const;
+
+  // TODO Implement specific cases where a leaf size can be selected.
```

Comment:
I asked the same thing. :)  https://github.com/mlpack/mlpack/pull/1301#discussion_r245723816  I think we'll open an issue after the merge is done and it can get handled later.

Reply:
I can handle the issue in the future :) The problem is I won't have time until mid-February probably